File: 07_LangChain - Agents/nodejs/currentWeatherTool.js
import { DynamicTool } from '@langchain/core/tools'
import dotenv from 'dotenv'
dotenv.config()

const WEATHER_API_KEY = process.env.WEATHER_API_KEY

const getCurrentWeatherTool = new DynamicTool({
	name: 'get current weather',
	description: 'Returns the current weather for a city,countryCode (e.g. Jerusalem,IL)',
	func: async (location) => {
		const regex = /^[A-Za-z\s]+,[A-Z]{2}$/
		if (!regex.test(location)) return 'Invalid format. Use City,CountryCode (e.g. Jerusalem,IL)'
		const params = new URLSearchParams({
			appid: WEATHER_API_KEY,
			units: 'metric',
			lang: 'en',
			q: location
		})
		const res = await fetch(`https://api.openweathermap.org/data/2.5/weather?${params}`)
		if (res.ok) {
			const data = await res.json()
			return `Temperature: ${data.main.temp}°C, Description: ${data.weather[0].description}`
		} else return 'Error fetching weather data.'
	}
})

export default getCurrentWeatherTool

File: 07_LangChain - Agents/python/current_weather_tool.py
from langchain_core.tools import tool
import requests, re
from secret_keys import open_weather_key


@tool
def get_current_weather(location) -> dict:
    """
    get current weather information for a given city, country code. The input must be in the format 'City,CountryCode'
    """
    pattern = r'^[^,]+,[A-Za-z]{2}$'
    if not re.match(pattern, location):
        return "error: Invalid input format. Expected 'City,CountryCode', e.g., 'Jerusalem,IL'"

    params = {
        "appid": open_weather_key,
        "units": "metric",
        "lang": "en",
        "q": location,
    }

    res = requests.get("https://api.openweathermap.org/data/2.5/weather", params=params)
    if res.status_code == 200:
        data = res.json()
        return f"temp: {data['main']['temp']}°C, description: {data['weather'][0]['description']}"
    else:
        return "error: Failed to fetch weather data"


if __name__ == "__main__":
    weather = get_current_weather.invoke(
        input={"location": "Tel Aviv,IL"}
    )
    print(f"{weather['temp']}°C, {weather['description']}")
